code/campaignclassic/src/main/java/com/adobe/marketing/mobile/FilmstripCarouselTemplateNotificationBuilder.java

```diff
@@ -0,0 +1,363 @@
+                            + " constructed.");
+        }
+
+        // download the carousel images and populate the image uri, image caption, and image click
+        // action arrays
+        final int centerImageIndex =
+                CampaignPushConstants.DefaultValues.CENTER_INDEX; // center index defaults to 1
+        final long imageProcessingStartTime = System.currentTimeMillis();
+        final List<CarouselPushTemplate.CarouselItem> items = pushTemplate.getCarouselItems();
+        final ArrayList<Bitmap> downloadedImages = new ArrayList<>();
+        final ArrayList<String> downloadedImageUris = new ArrayList<>();
+        final ArrayList<String> imageCaptions = new ArrayList<>();
+        final ArrayList<String> imageClickActions = new ArrayList<>();
+
+        for (final CarouselPushTemplate.CarouselItem item : items) {
+            final String imageUri = item.getImageUri();
+            final Bitmap pushImage = CampaignPushUtils.downloadImage(cacheService, imageUri);
+            if (pushImage != null) {
+                downloadedImages.add(pushImage);
+                downloadedImageUris.add(imageUri);
+                imageCaptions.add(item.getCaptionText());
+                imageClickActions.add(item.getInteractionUri());
+            }
+        }
+
+        // log time needed to process the carousel images
+        final long imageProcessingElapsedTime =
+                System.currentTimeMillis() - imageProcessingStartTime;
+        Log.trace(
+                CampaignPushConstants.LOG_TAG,
+                SELF_TAG,
+                "Processed %d manual filmstrip carousel image(s) in %d milliseconds.",
+                downloadedImageUris.size(),
+                imageProcessingElapsedTime);
+
+        // fallback to a basic push template notification builder if less than 3 images were able
+        // to be downloaded
+        if (downloadedImageUris.size()
+                < CampaignPushConstants.DefaultValues.FILMSTRIP_CAROUSEL_MINIMUM_IMAGE_COUNT) {
+            Log.trace(
+                    CampaignPushConstants.LOG_TAG,
+                    SELF_TAG,
+                    "Only %d image(s) for the manual filmstrip carousel notification were"
+                            + " downloaded. Building a basic push notification instead.",
+                    downloadedImageUris.size());
+            if (!CollectionUtils.isEmpty(downloadedImageUris)) {
+                // use the first downloaded image (if available) for the basic template notification
+                pushTemplate.modifyData(
+                        CampaignPushConstants.PushPayloadKeys.IMAGE_URL,
+                        downloadedImageUris.get(0));
+            }
+            final BasicPushTemplate basicPushTemplate =
+                    new BasicPushTemplate(pushTemplate.getData());
+            return BasicTemplateNotificationBuilder.construct(basicPushTemplate, context);
+        }
```

Comment:
Can you extract this into a separate private function within this class.

Reply:
i've moved this to a function named `fallbackToBasicNotification` located in the `CarouselTemplateNotificationBuilder` class as both auto and manual carousel notifications will use it.